doc/reference/guidelines.rst

```diff
@@ -0,0 +1,424 @@
+    import base64
+    import re
+    import time
+    # 2 :  imports of openerp
+    import openerp
+    from openerp import api, fields, models # alphabetically ordered
+    from openerp.tools.safe_eval import safe_eval as eval
+    from openerp.tools.translate import _
+    # 3 :  imports from odoo modules
+    from openerp.addons.website.models.website import slug
+    from openerp.addons.web.controllers.main import login_redirect
+
+
+Idioms
+-------
+
+- Prefer ``%`` over ``.format()``, prefer ``%(varname)`` instead of position (This is better for translation)
+- Try to avoid generators and decorators
+- Always favor *Readability* over *conciseness* or using the language features or idioms.
+- Use list comprehension, dict comprehension, and basic manipulation using ``map``, ``filter``, ``sum``, ... They make the code easier to read.
+- The same applies for recordset methods : use ``filtered``, ``mapped``, ``sorted``, ...
+- Each python file should have ``# -*- coding: utf-8 -*-`` as first line
+- Use the ``UserError`` defined in ``openerp.exceptions`` instead of overriding ``Warning``, or find a more appropriate exception in *exceptions.py*
+- Document your code (docstring on methods, simple comments for the tricky part of the code)
+- Use meaningful variable/class/method names
+
+
+
+Symbols
+-------
+
+- Odoo Python Class : use camelcase for code in api v8, underscore lowercase notation for old api.
+
+.. code-block:: python
+
+    class AccountInvoice(models.Model):
+        ...
+
+    class account_invoice(osv.osv):
+        ...
+
+- Variable name :
+    - use camelcase for model variable
+    - use underscore lowercase notation for common variable.
+    - since new API works with record or recordset instead of id list, don't suffix variable name with *_id* or *_ids* if they not contain id or list of id.
+
+.. code-block:: python
+
+    ResPartner = self.env['res.partner']
+    partners = ResPartner.browse(ids)
+    partner_id = partners[0].id
+
+- ``One2Many`` and ``Many2Many`` fields should always have *_ids* as suffix (example: sale_order_line_ids)
+- ``Many2One`` fields should have *_id* as suffix (example : partner_id, user_id, ...)
+- Method conventions
+    - Compute Field : the compute method pattern is *_compute_<field_name>*
+    - Search method : the search method pattern is *_search_<field_name>*
+    - Default method : the default method pattern is *_default_<field_name>*
+    - Onchange method : the onchange method pattern is *_onchange_<field_name>*
+    - Constraint method : the constraint method pattern is *_check_<constraint_name>*
+    - Action method : an object action method is prefix with *action_*. Its decorator is ``@api.multi``, but since it use only one record, add ``self.ensure_one()`` at the beginning of the method.
+
+- In a Model attribute order should be
+    #. Private attributes (``_name``, ``_description``, ``_inherit``, ...)
+    #. Default method and ``_default_get``
+    #. Fields declarations
```

Comment:
Default have to be defined before fields, because we give fields function pointers, not strings.